iroh-net/src/magic_endpoint.rs

```diff
@@ -560,10 +560,40 @@ impl MagicEndpoint {
     ///
     /// Returns an error if closing the magic socket failed.
     /// TODO: Document error cases.
-    pub async fn close(&self, error_code: VarInt, reason: &[u8]) -> Result<()> {
-        self.cancel_token.cancel();
-        self.endpoint.close(error_code, reason);
-        self.msock.close().await?;
+    pub async fn close(self, error_code: VarInt, reason: &[u8]) -> Result<()> {
+        let MagicEndpoint {
+            msock,
+            endpoint,
+            cancel_token,
+            ..
+        } = self;
+        cancel_token.cancel();
+        endpoint.close(error_code, reason);
+
+        {
+            // waiting for connections to close can be slow: this timeout decide whethers we should
+            // inform the user of a slow operation
+            let logging_timeout = tokio::time::sleep(Duration::from_millis(500));
+            futures::pin_mut!(logging_timeout);
+
+            let wait_idle = endpoint.wait_idle();
+            futures::pin_mut!(wait_idle);
+
+            // race the timeout and closing. Nothing else needs to be done if closing finishes
+            // first. If the timeouts is done first, give the user some logs
+            if let future::Either::Right(((), wait_idle_fut)) =
+                future::select(wait_idle, logging_timeout).await
+            {
+                tracing::info!("Closing connections");
+                wait_idle_fut.await;
```

Comment:
when testing, did you observe this often? or is this more a rare thing that it takes longer than 500ms?